### PR TITLE
refs #76 add scan range support

### PR DESCRIPTION
**Pull request by vyacheslavs**
range scan support added to scanmem console. No changes made to gui

### CODE REVIEW

File: handlers.c [hunk 1]
```diff
@@ -745,17 +745,25 @@ bool handler__lregions(globals_t * vars, char **argv, unsigned argc)
     return true;
 }
 
+static bool do_search ( scan_match_type_t m, globals_t * vars, uservalue_t * pval, char ** argv, unsigned argc );
```
Comment (sriemer) [inline handlers.c]: There are style issues in your commits. Don't put any spaces between function name and '(', don't put any spaces behind '(', don't put any spaces between '\*' and variable name and don't put any spaces in front of ')'.
I know there are some legacy parts which have that space between '\*' and the variable name but we're gradually fixing that when we change something at these locations.

So this line should look like this:
```C
static bool do_search(scan_match_type_t m, globals_t *vars, uservalue_t *val, char **argv, unsigned argc);
```
But using forward declaration for static functions is cumbersome as two locations have to be changed if something changes. Please use bottom-up to avoid forward declaration.

File: handlers.c [hunk 2]
```diff
@@ -745,17 +745,25 @@ bool handler__lregions(globals_t * vars, char **argv, unsigned argc)
...
+static bool try_get_range_values ( const char * arg, uservalue_t ** pval );
+
 /* the name of the function is for history reason, now GREATERTHAN & LESSTHAN are also handled by this function */
 bool handler__decinc(globals_t * vars, char **argv, unsigned argc)
 {
     uservalue_t val;
-    scan_match_type_t m;
+    uservalue_t * pval = &val;
```
Comment (sriemer) [inline handlers.c]: Don't introduce new pointer variables for an existing variable which is used in a similar way at other locations if not changing that everywhere. That `uservalue_t val` stuff is used at multiple places, you are touching unrelated code parts, so please don't introduce that `pval` variable.

File: handlers.c [hunk 3]
```diff
@@ -745,17 +745,25 @@ bool handler__lregions(globals_t * vars, char **argv, unsigned argc)
...
+    scan_match_type_t m = MATCHUNSPECIFIED;
 
     USEPARAMS();
 
     if (argc == 1)
     {
-        zero_uservalue(&val);
+        zero_uservalue(pval);
+        if (try_get_range_values(argv[0], &pval))
```
Comment (sriemer) [inline handlers.c]: The intention is that K&R is used. So the brace belongs to the end of the `if` line - of cause with a space in front of it.

Looks like this:
```C
if (try_get_range_values(argv[0], &pval)) {
```

The only exception are the braces used for function definition.

There are legacy parts which don't use K&R like the `if (argc == 1)` above. We are gradually fixing that once we change something at these locations.

File: handlers.c [hunk 4]
```diff
@@ -795,27 +809,91 @@ bool handler__decinc(globals_t * vars, char **argv, unsigned argc)
...
     if (vars->matches) {
-        if (checkmatches(vars, m, &val) == false) {
+        if (checkmatches(vars, m, pval) == false) {
             show_error("failed to search target address space.\n");
             return false;
         }
     } else {
         /* < > = != cannot be the initial scan */
-        if (argc == 1)
+        if (argc == 1 && m!=MATCHRANGE)
```
Comment (sriemer) [inline handlers.c]: Here the spaces are missing. This is hard to read and you've completely mixed styles.
This should be `m != MATCHRANGE`.

File: scanroutines.c [hunk 5]
```diff
@@ -226,6 +227,48 @@ DEFINE_FLOAT_INCREASEDBY_DECREASEDBY_ROUTINE(FLOAT32, 32)
...
+           (get_s##DATAWIDTH##b(new_value)<=get_s##DATAWIDTH##b(&user_value[1])) ) { \
+            ret = (DATAWIDTH)/8; SET_FLAG(saveflags, s##DATAWIDTH##b); \
+        } \
+        if ((new_value->flags.u##DATAWIDTH##b) && (user_value[0].flags.u##DATAWIDTH##b) && \
+           (user_value[1].flags.u##DATAWIDTH##b) && (get_u##DATAWIDTH##b(new_value)>=get_u##DATAWIDTH##b(&user_value[0])) && \
+           (get_u##DATAWIDTH##b(new_value)<=get_u##DATAWIDTH##b(&user_value[1])) ) { \
+            ret = (DATAWIDTH)/8; SET_FLAG(saveflags, u##DATAWIDTH##b); \
+        } \
+        return ret; \
+    }
```
Comment (sriemer) [inline scanroutines.c]: Better rethink this part. This is unreadable.

File: handlers.c [hunk 6]
```diff
@@ -795,27 +809,91 @@ bool handler__decinc(globals_t * vars, char **argv, unsigned argc)
...
+    if (!argv_copy)
+        return false;
+
+    char * token = strtok(argv_copy, "..");
+
+    uservalue_t * pval = NULL;
+
+    if (pval_out)
+    {
+        pval = malloc(sizeof(uservalue_t)*2);
```
Comment (sriemer) [inline handlers.c]: Better clearly show that this is multiplication. Put spaces before and after '\*'.

File: handlers.c (hunk 1 above)
Comment (vyacheslavs) [inline handlers.c]: could you please tell me which coding style should i use, so i can setup my uncrustify utility. thanks a lot

Comment (sriemer) [inline handlers.c]: Linux kernel coding style with the only exception not to use tabs but 4 spaces for indentation instead.
See: https://www.kernel.org/doc/Documentation/CodingStyle

**Comment by sriemer** [comment]
Thanks for reworking. The GUI part should be split from the rest into a separate commit. Also a long description is mandatory - especially for bigger changes. Imagine that you would have to review this.

**Comment by sriemer** [comment]
I guess as undocumented this is, as untested this is.
I have a small float test like this compiled with -O0:
````
#include <stdio.h>
#include <unistd.h>

static float magic1 = 1.34;
static float magic2 = 1.35;
static float magic3 = 1.36;

int main(void)
{
        while (1) {
                sleep(1);
                printf("%f, %f, %f\n", magic1, magic2, magic3);
        }
        return 0;
}
````
I've tried searching for `1.34..1.35` with the scan data type float and I get 180 matches with matches like "32.000000". This is definitely not working properly.

For what does this stuff work?
Why is the "<>" command using the decinc handler required if using the default handler otherwise?

**Comment by vyacheslavs** [comment]
there was no updates to this topic for so much time, so i thought it was abandoned. I need some time to rework patches, cause now I am too busy. Guess I need some time also to remember the code, cause it was long ago. Please bear with it.

**Comment by sriemer** [comment]
I will implement this stuff. But this really shows how important it is to split stuff into atomic changes, describe changes, to test stuff and not to introduce a new style. I also maintain this in my free time and bad commits are causing huge delays or even rejection as nobody wants to review or reverse engineer that. Often it is faster to drop such stuff and to implement it from scratch like it should be.
Review should not only be the task of the maintainers. Also watchers are very welcome to review changes as they should have an interest that the tool they watch doesn't become broken.

**Comment by sriemer** [comment]
Implemented differently. So closing.
